File: src/algorithms/gcm.jl

module GCM


include("../util/semantic_types.jl")
using .SemanticTypes

include("sea128.jl")
include("../math/galois_fast.jl")
using Nettle
using Base64
using .Sea128: encrypt_sea, decrypt_sea
using .Galois_quick: FieldElement


arr_to_int(arr::Array{UInt8}) = reinterpret(UInt128, reverse(arr))[1]
padl(len::Int) = (16 - (len % 16)) % 16 ## pad with zeros to the next multiple of 16
pad_array(arr::Array{UInt8}) = [arr; zeros(UInt8, padl(length(arr)))]

function ghash(key::Array{UInt8}, nonce::Array{UInt8}, text::Array{UInt8}, ad::Array{UInt8}, algorithm::String)

    enc_func = algorithm == "aes128" ? encrypt : encrypt_sea 
    auth_key = enc_func("aes128", key, zeros(UInt8, 16))
    auth_key = FieldElement(arr_to_int(auth_key), SemanticTypes.GCM)

    len_block = vcat(
        reverse(reinterpret(UInt8, [length(ad) << 3])),
        reverse(reinterpret(UInt8, [length(text) << 3]))
    )
    
    Y = FieldElement(UInt128(0), SemanticTypes.GCM)
    data = [pad_array(ad); pad_array(text); len_block]

    for i in 1:16:(length(data) - 1)
        Y += data[i:i+15]
        Y *= auth_key
    end

    tag = Y + enc_func("AES128", key, [nonce; UInt8[0,0,0,1]])

    return tag, len_block, auth_key
end


function crypt_gcm(key::Array{UInt8}, text::Array{UInt8}, nonce::Array{UInt8}, algorithm::String)

    result_text = Array{UInt8}(undef, 0)
    enc_func = algorithm == "aes128" ? encrypt : encrypt_sea


    counter = 2
    for i in 1:16:(length(text))
        temp_nonce = [nonce; reverse!(reinterpret(UInt8, [UInt32(counter)]))]
        enc_i = enc_func("AES128", key, temp_nonce)
        end_idx = min(i+15, length(text))
        enc_block = enc_i[1:length(block)]
        append!(result_text, text[i:end_idx] .⊻ enc_block)
        counter += 1
    end

    return result_text
end

function decrypt_gcm(key::Array{UInt8}, ciphertext::Array{UInt8}, ad::Array{UInt8}, nonce::Array{UInt8}, algorithm::String)
    auth_tag = ghash(key, nonce, ciphertext, ad, algorithm)
    plaintext = crypt_gcm(key, ciphertext, nonce, algorithm)
    return plaintext, auth_tag[1], auth_tag[2], auth_tag[3]
end


function encrypt_gcm(key::Array{UInt8}, plaintext::Array{UInt8}, ad::Array{UInt8}, nonce::Array{UInt8}, algorithm::String)
    ciphertext = crypt_gcm(key, plaintext, nonce, algorithm)
    auth_tag = ghash(key, nonce, ciphertext, ad, algorithm)
    return ciphertext, auth_tag[1], auth_tag[2], auth_tag[3]
end


end

File: src/algorithms/gcm.cpp
// src/algorithms/gcm.cpp
#include "gcm.h"
#include "../util/semantic_types.h"
#include "../util/helper_functions.h"
#include "../math/galois.h"
#include "../algorithms/sea128.h"
#include <vector>
#include <cstdint>
#include <stdexcept>
#include <tuple>
#include <algorithm>

#include <functional>


std::vector<uint8_t> pad_array(const std::vector<uint8_t>& arr) {
    size_t pad_len = (16 - (arr.size() % 16)) % 16;
    std::vector<uint8_t> padded = arr;
    padded.resize(arr.size() + pad_len, 0);
    return padded;
}


std::tuple<FieldElement, std::vector<uint8_t>, FieldElement> ghash(
    const std::vector<uint8_t>& key,
    const std::vector<uint8_t>& nonce,
    const std::vector<uint8_t>& text,
    const std::vector<uint8_t>& ad,
    const std::string& algorithm) {

    std::function<std::vector<uint8_t>(const std::vector<uint8_t>&, const std::vector<uint8_t>&)> enc_func;
    enc_func = algorithm == "aes128" ? aes_encrypt : encrypt_sea;

    std::vector<uint8_t> zero_block(16, 0);
    std::vector<uint8_t> encrypted_zero = enc_func(key, zero_block);
    __uint128_t auth_key_int = bytes_to_uint128(encrypted_zero);
    FieldElement auth_key(auth_key_int, Semantic::GCM, false);

    uint64_t ad_bit_len = static_cast<uint64_t>(ad.size()) * 8;
    uint64_t text_bit_len = static_cast<uint64_t>(text.size()) * 8;

    // Convert to big endian bytes
    std::vector<uint8_t> len_block(16, 0);
    for(int i = 0; i < 8; ++i){
        len_block[15 - i] = (text_bit_len >> (i * 8)) & 0xFF;
        len_block[7 - i] = (ad_bit_len >> (i * 8)) & 0xFF;
    }

    std::vector<uint8_t> padded_ad = pad_array(ad);
    std::vector<uint8_t> padded_text = pad_array(text);


    // Concatenate padded AD, padded text, and len_block
    std::vector<uint8_t> data;
    data.reserve(padded_ad.size() + padded_text.size() + len_block.size());
    data.insert(data.end(), padded_ad.begin(), padded_ad.end());
    data.insert(data.end(), padded_text.begin(), padded_text.end());
    data.insert(data.end(), len_block.begin(), len_block.end());

    // Initialize Y as zero
    FieldElement Y(0, Semantic::GCM, false);

    // Iterate over each 16-byte block
    for(size_t i = 0; i < data.size(); i += 16){
        std::vector<uint8_t> block(data.begin() + i, data.begin() + i + 16);
        __uint128_t block_int = bytes_to_uint128(block);
        FieldElement block_fe(block_int, Semantic::GCM, false);
        Y = Y + block_fe;
        Y = Y * auth_key;
    }

    std::vector<uint8_t> ctr_block = nonce;
    if(nonce.size() != 12){
        throw std::invalid_argument("Nonce must be 12 bytes for GCM.");
    }
    ctr_block.insert(ctr_block.end(), {0, 0, 0, 1}); // TODO: Fix this
    std::vector<uint8_t> S = enc_func(key, ctr_block);

    // Convert tag_int to bytes
    FieldElement tag = Y + S;

    return {tag, len_block, auth_key};
}

std::vector<uint8_t> crypt_gcm(
        const std::vector<uint8_t>& key,
        const std::vector<uint8_t>& text,
        const std::vector<uint8_t>& nonce,
        const std::string& algorithm) {
    std::vector<uint8_t> result_text;


    std::function<std::vector<uint8_t>(const std::vector<uint8_t>&, const std::vector<uint8_t>&)> enc_func;
    enc_func = algorithm == "aes128" ? aes_encrypt : encrypt_sea;

    uint32_t counter = 2;
    for (size_t i = 0; i < text.size(); i += 16) {
        std::vector<uint8_t> temp_nonce = nonce;
        temp_nonce.push_back((counter >> 24) & 0xFF);
        temp_nonce.push_back((counter >> 16) & 0xFF);
        temp_nonce.push_back((counter >> 8) & 0xFF);
        temp_nonce.push_back(counter & 0xFF);
        std::vector<uint8_t> enc_ctr = enc_func(key, temp_nonce);

        std::vector<uint8_t> block(text.begin() + i, text.begin() + std::min(i + 16, text.size()));
        std::vector<uint8_t> encrypted = enc_func(key, block);
        // XOR with encrypted counter block
        for (size_t j = 0; j < encrypted.size(); ++j) {
            encrypted[j] = enc_ctr[j] ^ block[j];
        }
        result_text.insert(result_text.end(), encrypted.begin(), encrypted.end());
        counter++;
    }

    return result_text;
}


std::tuple<std::vector<uint8_t>, FieldElement, std::vector<uint8_t>, FieldElement> encrypt_gcm(
        const std::vector<uint8_t>& key,
        const std::vector<uint8_t>& plaintext,
        const std::vector<uint8_t>& ad,
        const std::vector<uint8_t>& nonce,
        const std::string& algorithm) {

    auto cipertext = crypt_gcm(key, plaintext, nonce, algorithm);
    auto [auth_tag, len_block, auth_key] = ghash(key, nonce, cipertext, ad, algorithm);

    return {cipertext, auth_tag, len_block, auth_key};
}


std::tuple<std::vector<uint8_t>, std::vector<uint8_t>, std::vector<uint8_t>, std::vector<uint8_t>> decrypt_gcm(
        const std::vector<uint8_t>& key,
        const std::vector<uint8_t>& ciphertext,
        const std::vector<uint8_t>& ad,
        const std::vector<uint8_t>& nonce,
        const std::string& algorithm) {

    auto [auth_tag, len_block, auth_key] = ghash(key, nonce, ciphertext, ad, algorithm);
    auto plaintext = crypt_gcm(key, ciphertext, nonce, algorithm);

    return {plaintext, auth_tag.to_vector(), len_block, auth_key.to_vector()};
}